### ISSUE TITLE
Yolo v2 tiny voc results are different in ONNX as that of results from KERAS

### ISSUE THREAD

**Issue by iamshri8** [onnx/keras-onnx#84]
Hello,

I converted yolo v2 tiny voc model from darknet to keras using [yad2k script](https://github.com/allanzelener/YAD2K) and tested the model with the following image:
![dog-pund](https://user-images.githubusercontent.com/31486526/57395557-ab216080-71c8-11e9-92f5-1a4c5c58de0f.jpeg)
 
Below is the result for the above image on yolo v2 tiny voc:

![scores](https://user-images.githubusercontent.com/31486526/57395664-fb98be00-71c8-11e9-96a5-7c2db41a512a.png)

And also below is the first 10 values of the network output in keras:
![Screen Shot 2019-05-08 at 7 25 17 PM](https://user-images.githubusercontent.com/31486526/57395715-1cf9aa00-71c9-11e9-8625-2bb1e04d0b2a.png)

The converted keras model is as follows:
[keras model](https://drive.google.com/open?id=1XqU2ZoycMghASswD-USVMwQQr3y75P38)

I used your script to convert the above keras model into Onnx( using channel_first input).

**Now the final results and the output values of the converted onnx model is different from keras model.** Please find the results of the onnx model below:

![image](https://user-images.githubusercontent.com/31486526/57396201-26cfdd00-71ca-11e9-83d1-fe9840269d6a.png)

And the output values of the onnx model are as follows:
![image](https://user-images.githubusercontent.com/31486526/57396261-5383f480-71ca-11e9-86b4-9de442922079.png)

Please find the onnx model below:
[Onnx model - yolo](https://drive.google.com/open?id=1-ikACFWO8RrnwtDa9OljNqVt5PElWgMF)

Could you please help me in this?

Thanks for your help in advance.

**Comment by jiafatom** [comment]
@iamshri8 Can you compare the output for those inner nodes to see if it matches? We need locate which node causes wrong result.

**Comment by iamshri8** [comment]
I compared some of the inner nodes from the beginning. The first conv2d layer has a slight difference in the results. So, it is carried over other layers and there is more difference in the output results.

**Please find the first 10 output results of first conv2d layer in keras and ONNX:**
_Keras:_
![subModelyolo_conv_2d_1](https://user-images.githubusercontent.com/31486526/57529959-70d5d180-7336-11e9-85fb-4f07bdf3c579.png)
_ONNX:_
![subModelYolo_conv2d_1](https://user-images.githubusercontent.com/31486526/57530994-c57a4c00-7338-11e9-8fa4-11048b26a1dd.png)

**Results of 1st batch normalization layer:**
_Keras:_
![subModelyolo_batch_normalization_1](https://user-images.githubusercontent.com/31486526/57530046-a975ab00-7336-11e9-937e-ecdf4446a0b6.png)
_ONNX:_
![subModelYolo_batch_normalization_1](https://user-images.githubusercontent.com/31486526/57531019-d1660e00-7338-11e9-8f55-97560636c865.png)

**Results after 5th Conv2d:**
_Keras:_
![subModelYolo_conv5](https://user-images.githubusercontent.com/31486526/57530096-cb6f2d80-7336-11e9-97e6-d34bd5a2163a.png)
_ONNX:_
![subModelYolo_conv2d_5](https://user-images.githubusercontent.com/31486526/57530871-80561a00-7338-11e9-8a98-b509a45662c3.png)

As we can see, the slight difference in output results in each layer, leads to wrong confidence scores in the output results.

**Comment by jiafatom** [comment]
Slight numerical difference is fine, because onnxruntime implementation may cause that. We usually treat the result is correct within 1% error. For the 1st conv2d, I feel the result is good.

[Here](https://github.com/onnx/keras-onnx/blob/master/tests/test_layers.py) contains example how to compare using run_onnx_runtime. You can test to see which op causes issue.

**Comment by iamshri8** [comment]
@jiafatom , Thanks for the reply. 

As I mentioned, results after 7th conv2d layer is totally different. Please find the attached results.

Keras:
![57533128-4b989180-733d-11e9-90b5-dda150594398](https://user-images.githubusercontent.com/31486526/57553310-77803b00-736e-11e9-8b80-3859aafbebe9.png)
ONNX results:
![subModelYolo_conv2d_7](https://user-images.githubusercontent.com/31486526/57553311-7818d180-736e-11e9-8236-2a6bdaa7b45a.PNG)

As we see the first 4 values are negative in Keras whereas the values are not negative in ONNX. Please see the attached images.

I found out that error is at 7th conv2d layer. So, is there a way i can get correct results?

**Comment by jiafatom** [comment]
We expect to get the layer which satisfies (1) this layer get wrong results (2) its previous layer get correct result (using 1% error criteria). Then we can say this layer has problem. Is this 7th conv2d layer the one that satisfies these two? If it is, one thing you could do is to check the parameters and weights in the layer to see if it is correctly set up in onnx model.

**Comment by iamshri8** [comment]
@jiafatom 
Thanks for the reply.
I checked the weights in that layer. It seems to be the same in keras as well as onnx. But coulp you please tell me how to check the parameters?

**Comment by iamshri8** [comment]
@jiafatom 
Thanks for the reply.
I checked the weights in that layer. It seems to be the same in keras as well as onnx. But coulp you please tell me how to check the parameters?

**Comment by jiafatom** [comment]
Did you check the attributes in that layer? Compare onnx/Operators.md and keras conv2d to see if it is implemented in the same way.

**Comment by iamshri8** [comment]
@jiafatom , Yes i compared the attributes in keras conv2d and onnx/operators.md . Its implemented the same way. 
So, what could be the solution for this?


And also I tried converting other yolov2 and yolov3 models into ONNX, I get wrong results in all the models.

**Comment by jiafatom** [comment]
For yolov3,
Please do not use the code in yolov3.py, because there are some changes that are not checked-in.
Use the pretrained yolov3 onnx model [here](https://github.com/onnx/models/tree/master/yolov3).

**Comment by iamshri8** [comment]
@jiafatom , Ok but i tried loading this model. But unable to load the model. 

And what could be the solution for errors in other yolov2 models?

**Comment by jiafatom** [comment]
Do you mean that you cannot run yolov3 on onnxruntime? What is the error message?
It is fine to see these isolated cast ops, sometimes it is because netron version. I use netron 3.0.5 it does not have these cast ops.

**Comment by iamshri8** [comment]
@jiafatom  , No I am not trying to load on onnxruntime. I am trying to load in windows ML and I get this error.

![Capture](https://user-images.githubusercontent.com/31486526/57799465-74f75a00-774f-11e9-945d-d4694f95dd1f.PNG)

**Comment by jiafatom** [comment]
yolov3 only has two inputs, why we have input size 5?

**Comment by iamshri8** [comment]
@jiafatom , I am not sure why this error. But When i tried converting the model using yolov3.py script the yolov3 model looked like below.

![Capture](https://user-images.githubusercontent.com/31486526/57800324-71fd6900-7751-11e9-9d45-68d3cbc6823b.PNG)


This also has 2 inputs , I am able to load this. But I am unable to load the model from onnx zoo.

**Comment by jiafatom** [comment]
Please don't try to convert yolov3 by yourself, because there is some code that are not checked-in.

For model zoo yolov3, when you open netron, how many inputs it expects?

**Comment by iamshri8** [comment]
@jiafatom  , Its just height and width so I think its 2.

**Comment by iamshri8** [comment]
@jiafatom , Sorry for the above comment. Just got confused. It takes 2 inputs. One is image and one more is input shape

**Comment by jiafatom** [comment]
For yolov3 onnx, we tested via onnxruntime successfully. When you mentioned input size 5 in windows ML, do you know which op it is? (as you mentioned you just have 2 inputs, right?)

**Comment by jiafatom** [comment]
@iamshri8 , windows ML only supports opset 8 or earlier, but yolov3 need opset 10. So this model cannot run on windows ML.

**Comment by iamshri8** [comment]
@jiafatom , Thanks for the update .

**Comment by iamshri8** [comment]
@jiafatom , And what about yolov2 models?

Is there a way to resolve this ?

**Comment by jiafatom** [comment]
@iamshri8 I just get chance looking at your yolov2 model. Actually I made several fixes recently on batch_normalization, upsample, etc. When I convert your yolov2 model today, and find that the end-to-end result matches well. Can you please pull latest keras-onnx master and convert your model again? Thanks for your patience.

**Comment by iamshri8** [comment]
@jiafatom , Thank you for the update. I uninstalled the keras2onnx and installed the latest version keras2onnx 1.5.0 using pip and converted the yolov2 model. I still get different results when i load the onnx model in windows ml to the keras results :(

**Comment by jiafatom** [comment]
@iamshri8 The current issue is on windows ml side. The engine of windows ml is onnxruntime but it only supports until opset 7 or 8. However, current onnxruntime supports opset 10. This means that windows ml does not pick the latest onnxruntime bug fixes, etc. I ran the model correctly on latest onnxruntime master, so this is not the converter issue any more.

**Comment by iamshri8** [comment]
@jiafatom  , Thank you so much for the support!!!!
